File: app/src/main/java/com/example/saka/myapplication/RegisterActivity.java
package com.example.saka.myapplication;

import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.example.saka.myapplication.model.User;
import com.example.saka.myapplication.model.myFirebase;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

public class RegisterActivity extends AppCompatActivity {

    private EditText edtName;
    private EditText edtEmail;
    private EditText edtPassword;
    private FirebaseAuth auth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null){
            actionBar.hide();
        }

        edtName = (EditText) findViewById(R.id.su_edt_name);
        edtEmail = (EditText) findViewById(R.id.su_edt_mail);
        edtPassword = (EditText) findViewById(R.id.su_edt_password);
        Button btnSignUp = (Button) findViewById(R.id.su_btn_sign_up);
        TextView txtLogIn = (TextView) findViewById(R.id.su_txt_login);

        auth = FirebaseAuth.getInstance();

        btnSignUp.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String enterEmail = edtEmail.getText().toString();
                String enterPassword= edtPassword.getText().toString();
                createNewUser(enterEmail, enterPassword);
            }
        });

        txtLogIn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), LogInActivity.class));
                finish();
            }
        });
    }

    @Override
    protected void onStart() {
        super.onStart();
        if (auth.getCurrentUser() != null){
            onAuthSuccess(auth.getCurrentUser());
        }
    }

    private void createNewUser(String email, String password){
        if (!validateForm()){
            return;
        }
        auth.createUserWithEmailAndPassword(email, password)
                .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (!task.isSuccessful()){
                            Toast.makeText(getApplicationContext(), "Create account failed", Toast.LENGTH_SHORT).show();
                        } else {
                            onAuthSuccess(task.getResult().getUser());
                        }
                    }
                });
    }

    private void onAuthSuccess(FirebaseUser user){
        // write new user
        String username = edtName.getText().toString();
        User newUser = new User(user.getUid(), username, user.getEmail());
        myFirebase.writeNewUser(user.getUid(), newUser);
        Log.d("REGISTER", "write new user success");

        startActivity(new Intent(getApplicationContext(), MainActivity.class)
                .putExtra("name", edtName.getText().toString()));
        finish();
    }

    private boolean validateForm(){
        boolean valid = true;
        //Validate Password
        String password = edtPassword.getText().toString();
        if (TextUtils.isEmpty(password)){
            edtPassword.setError("Create your new password");
            valid = false;
        } else {
            edtPassword.setError(null);
        }
        //Validate Email
        String phone = edtEmail.getText().toString();
        if (TextUtils.isEmpty(phone)){
            edtEmail.setError("Enter your email address");
            valid = false;
        } else {
            edtEmail.setError(null);
        }
        //Validate Email
        String name = edtName.getText().toString();
        if (TextUtils.isEmpty(name)){
            edtName.setError("Enter your real name");
            valid = false;
        } else {
            edtName.setError(null);
        }
        return valid;
    }
}

File: app/src/main/java/com/example/saka/myapplication/ui/RegisterActivity.kt
package com.example.saka.myapplication.ui

import android.content.Intent
import android.content.pm.ActivityInfo
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.view.Window
import android.view.WindowManager
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.saka.myapplication.R
import com.example.saka.myapplication.model.MyFirebase.writeNewUser
import com.example.saka.myapplication.model.User
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser

class RegisterActivity : AppCompatActivity() {
    private var edtName: EditText? = null
    private var edtEmail: EditText? = null
    private var edtPassword: EditText? = null
    private var auth: FirebaseAuth? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        window.setFlags(
            WindowManager.LayoutParams.FLAG_FULLSCREEN,
            WindowManager.LayoutParams.FLAG_FULLSCREEN,
        )
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        val actionBar = supportActionBar
        actionBar?.hide()
        edtName = findViewById<View>(R.id.su_edt_name) as EditText
        edtEmail = findViewById<View>(R.id.su_edt_mail) as EditText
        edtPassword = findViewById<View>(R.id.su_edt_password) as EditText
        val btnSignUp = findViewById<View>(R.id.su_btn_sign_up) as Button
        val txtLogIn = findViewById<View>(R.id.su_txt_login) as TextView
        auth = FirebaseAuth.getInstance()
        btnSignUp.setOnClickListener {
            val enterEmail = edtEmail!!.text.toString()
            val enterPassword = edtPassword!!.text.toString()
            createNewUser(enterEmail, enterPassword)
        }
        txtLogIn.setOnClickListener {
            startActivity(Intent(applicationContext, LogInActivity::class.java))
            finish()
        }
    }

    override fun onStart() {
        super.onStart()
        if (auth!!.currentUser != null) {
            onAuthSuccess(auth!!.currentUser)
        }
    }

    private fun createNewUser(email: String, password: String) {
        if (!validateForm()) {
            return
        }
        auth!!.createUserWithEmailAndPassword(email, password)
            .addOnCompleteListener(this) { task ->
                if (!task.isSuccessful) {
                    Toast.makeText(applicationContext, "Create account failed", Toast.LENGTH_SHORT)
                        .show()
                } else {
                    onAuthSuccess(task.result.user)
                }
            }
    }

    private fun onAuthSuccess(user: FirebaseUser?) {
        // write new user
        val username = edtName!!.text.toString()
        val newUser = User(
            user!!.uid,
            username,
            user.email,
        )
        writeNewUser(user.uid, newUser)
        Log.d("REGISTER", "write new user success")
        startActivity(
            Intent(applicationContext, MainActivity::class.java)
                .putExtra("name", edtName!!.text.toString()),
        )
        finish()
    }

    private fun validateForm(): Boolean {
        var valid = true
        // Validate Password
        val password = edtPassword!!.text.toString()
        if (TextUtils.isEmpty(password)) {
            edtPassword!!.error = "Create your new password"
            valid = false
        } else {
            edtPassword!!.error = null
        }
        // Validate Email
        val phone = edtEmail!!.text.toString()
        if (TextUtils.isEmpty(phone)) {
            edtEmail!!.error = "Enter your email address"
            valid = false
        } else {
            edtEmail!!.error = null
        }
        // Validate Email
        val name = edtName!!.text.toString()
        if (TextUtils.isEmpty(name)) {
            edtName!!.error = "Enter your real name"
            valid = false
        } else {
            edtName!!.error = null
        }
        return valid
    }
}
